ReadMe
OFFICIAL RELEASE 2/16/12!

About: GetPerms is a plugin for Minecraft servers running Bukkit. This plugin will gather a list of all permission nodes for every plugin installed on the server and output the nodes to a file.

Installation:
  1. Download the plugin (you should have already)
  2. Move the jar file to '/plugins' inside the server root directory

Usage:
  1. Start the server
  2. Go into the server root directory
  3. Open 'pnodes.txt'

pnodes.txt:
  Each line is one separate permission node. Plugins must be in the '/plugins' folder and have permission nodes in '[plugin].jar/plugin.yml' in order for GetPerms to get the permission nodes from the plugin. Permission nodes are listed by plugins, which are in alphabetical order. A permission node's description is available in 'pnodesfull.txt' which is also in the server root directory. Useful if you want one to just copy and paste the nodes without having to highlight each line separately to avoid copying the description.

pnodesfull.txt
  Quite the same as pnodes.txt, except this one has the description next to each node.

Changelog.txt
Changelog:

v1.1.16 (1/21/16)
Fix incorrect version
Improved version comparison
Renamed package to use lowercase names

v1.1.15 (12/10/15)
Updates are no longer downloaded if it has already been downloaded and matches the checksum
Update command now displays output to the running player
Extra plugin-specific permissions now download from the plugin permission "database" when applicable
Updater retries under specific failure conditions (Curse ServerMods API returns empty JSON array or bad HTTP response code)

v1.1.14 (11/25/15)
Fixed the updater creating a folder instead of a file for updates
Refractored and cleaned up the update system
Added an update command, /getperms update
Update system now uses the Curse ServerMods API for release files
Changed update system to verify update files based on MD5 checksum instead of specific HTML elements (release files only)

v1.1.13 (11/24/15)
Fixed plugins.yml not saving
Added another file to cleanup

v1.1.12 (11/24/15)
Cleaned up a lot more code
Reduced redundant code
Reduced console output message count
Fixed always regenerating files when regenerateOnPluginChange was set to true
Improved command handling
Changed initial permission list generation delay from 20 seconds to 5 seconds
Changed configuration option autoGen to autoGenerate
Changed default configuration to make default behavior generate only when a plugin changes (autoGenerate changed to false)
pnodes.txt and pnodesfull.txt renamed to permission_nodes.txt and permission_nodes_desc.txt, respectively
Permission list files are now output to the plugin folder instead of the server directory
Plugin update jars are now downloaded to the plugin folder instead of update/
Automatically cleans up old files left from file name changes

v1.1.11 (11/23/15)
Updated metrics from R6 to R7

v1.1.10 (11/14/15)
Cleaned up and code and slightly changed coding style/convention

v1.1.9 (5/6/13)
Added update download validity checker (checks for HTML elements in file, if found, invalid)

v1.1.8 (5/6/13)
Added check to prevent double file generation at startup
Re-worded configuration file key descriptions
Moved redundant code to separate methods (changelog and readme downloaders)
New shiny command messages (for /getperms and /getperms help)
Moved redundant code to new class (command usage generator)
Changed one console message to be a one-time message only (less server message spam!)
Slightly modified command handling
Forgot to rename a permission node in plugin.yml (from copy and paste)

v1.1.7 (5/6/13)
Refractored file generation
Removed /gpregen and /getpermsregen commands
Added /getperms reload and /getperms help (you can still use /gp)
Re-did command help message
Optimized code
Official support for Bukkit 1.5.1-R0.2
ACTUALLY FIXED FileNotFoundException from v0.0.7

v1.1.6 (8/18/12)
Fixed regeneration on plugin change
Changed config handling
Bug: firstRun remains true
Bug: cfgV does not change

v1.1.5 (8/18/12) --Broken--
Added option in config to automatically regenerate permission files when a plugin is added/removed/updated
Slight code formatting
Slight code cleanup
Minor changes

v1.1.4 (5/10/12)
Fixed missing console output messages (Issue 8)

v1.1.3 (5/9/12)
Fixed NullPointerException (Plugin failed to load) (Issue 7)

v1.1.2 (5/7/12)
Fixed AccessControlException (Issue 6)
Attempt to fix the messages not being displayed in the console
Shortened a few method names

v1.1.1 (4/23/12)
Added temporary method to download plugin nodes from database
Shortened log output
Minor changes

v1.1.0 (4/21/12)
Minor change in command usage message

v1.0.9 (4/10/12)
Minor change in console output to make messages shorter
Added silentMode, control via config

v1.0.8 (3/24/12)
Fixed NPE in 1.0.7 (for real)
Minor changes

v1.0.7 (3/12/12) --Broken--
Fixed NPE in 1.0.6

v1.0.6 (3/8/12) --Broken--
Minor bugfixes
Removed unneeded code

v1.0.5 (3/6/12)
Added usage stat sending
Start of database retrieval
Minor changes here and there

v1.0.4 (3/5/12)
Removed StringBuilders
Cleaned code
Preparing to retrieve permission nodes from the database

v1.0.3 (3/1/12)
Fixed always downloading developmental build regardless of config

v1.0.2 (3/1/12)
Updated plugin to use new URL's; forgot to update in 1.0.1

v1.0.1 (3/1/12)
Change to method used to download files
Now downloads the latest recommended build, not the dev build
Added devBuilds key to config, to control downloading the latest devbuild or recommended build.
Actually fixed plugin always downloading ReadMe and Changelog files. The plugin is only supposed to download them when the plugin is updated.

v1.0.0 (2/28/12)
Major release
Added command to regenerate the permissions files, or generate it if it doesn't exist.
Useful if you set autoGen to false.
PermissionsEx support for the command (also supports whatever PEX supports)
Fixed plugin always downloading the ReadMe and Changelog files at each startup

v0.0.7 (2/21/12)
Added disable on finish in config
Control disable on finish in config
Added a class to create the popular comments in the config.yml
Minor logger message changes

v0.0.6 (2/21/12)
Added debug mode in config
Added auto-download in config
Control debug mode via config
Control auto-download via config

v0.0.5 (2/21/12)
Config is used
Control auto-update via config
Control auto-generate via config

v0.0.4 (2/21/12)
Major code change
Split code into multiple class files/methods
Added auto-update feature
Added config file (empty now)
Just wanted to get this out now
0.0.5 Coming out soon!

v0.0.3 (2/18/12)
Minor change for unnecessary newlines being generated

v0.0.2 (2/17/12)
Minor change to console output

v0.0.1 (2/16/11)
Initial release

src/com/github/GetPerms/GetPerms/Main.java
package com.github.getperms.getperms;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.net.MalformedURLException;
import java.security.DigestInputStream;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Set;
import java.util.logging.Logger;
import org.bukkit.Bukkit;
import org.bukkit.configuration.Configuration;
import org.bukkit.plugin.Plugin;
import org.bukkit.plugin.PluginDescriptionFile;
import org.bukkit.plugin.PluginManager;
import org.bukkit.plugin.java.JavaPlugin;
import org.bukkit.scheduler.BukkitScheduler;

public class Main extends JavaPlugin {

	public static final String PERMISSIONS_FILENAME = "permission_nodes.txt";
	public static final String PERMISSIONS_DESCRIPTION_FILENAME = "permission_nodes_desc.txt";

	public Configuration configuration;
	public ConfigHandler configHandler;
	private PluginManager pluginManager;
	private YamlHandler pluginListHandler;
	private BukkitScheduler scheduler;
	protected String pluginVersion;
	private File dataFolder;
	public File permissionFolder;
	protected File permissionNodes;
	protected File permissionNodesDesc;
	private File readMe;
	private File changeLog;
	public Logger logger;
	private boolean generationFinished = false;
	private int updateTaskId = -1;

	@Override
	public void onEnable() {
		configuration = getConfig();
		configuration.options().copyDefaults(true);
		saveConfig();

		pluginManager = Bukkit.getServer().getPluginManager();
		scheduler = Bukkit.getScheduler();

		configHandler = new ConfigHandler(this);
		configHandler.load();
		configuration = getConfig();
		logger = getLogger();
		debugValues();

		pluginListHandler = new YamlHandler(this, "plugins");

		PluginDescriptionFile pluginDescriptionFile = getDescription();
		pluginVersion = pluginDescriptionFile.getVersion();

		dataFolder = getDataFolder();
		permissionFolder = new File(dataFolder, "permissions");
		permissionNodes = new File(permissionFolder, PERMISSIONS_FILENAME);
		permissionNodesDesc = new File(permissionFolder, PERMISSIONS_DESCRIPTION_FILENAME);
		readMe = new File(dataFolder, "ReadMe.txt");
		changeLog = new File(dataFolder, "Changelog.txt");

		// Check and create the plugin data folder
		if (!dataFolder.exists()) {
			dataFolder.mkdirs();
		}

		if (!permissionFolder.exists()) {
			permissionFolder.mkdirs();
		}

		if (configuration.getBoolean("sendStats")) {
			info("Sending usage stats to mcstats.org every 15 minutes.");
			try {
				Metrics metrics = new Metrics(this);
				metrics.start();
			} catch (IOException e) {
				printStackTrace(e);
			}
		} else {
			info("The option to enable sending usage stats is in the configuration.");
		}
		debug("Configuration version: " + configuration.getString("cfgV") + " Plugin version: " + pluginVersion);
		if (!configuration.getString("cfgV").equalsIgnoreCase(pluginVersion)) {
			debug("Config version does not match jar version.");
			getStartFiles();
			cleanOldFiles();
		} else if (configuration.getBoolean("firstRun", true)) {
			debug("firstRun is set to true. Setting to false...");
			configuration.set("firstRun", false);
			getStartFiles();
		}
		configuration.set("cfgV", pluginVersion);

		configHandler.save();

		CommandHandler commandHandler = new CommandHandler(this);
		getCommand("getperms").setExecutor(commandHandler);
		getCommand("gp").setExecutor(commandHandler);

		if (configuration.getBoolean("autoUpdate", true)) {
			scheduler.scheduleSyncRepeatingTask(this, new UpdateTask(this), 10 * 20L, 60 * 60 * 24 * 20L);
		}

		if (configuration.getBoolean("regenerateOnPluginChange", true)) {
			debug("Checking for plugin changes...");

			if (!comparePlugins()) {
				debug("Changes found! Regenerating files...");
				generateFiles(true);
				generationFinished = true;
			}
		}

		writePluginList();

		if (configuration.getBoolean("autoGenerate", true)) {
			debug("autoGenerate enabled");
			generateFiles(true);
			generationFinished = true;
		}

		info("GetPerms " + pluginVersion + " enabled!");

		if (configuration.getBoolean("disableOnFinish", false)) {
			debug("DisableOnFinish enabled");
			info("Finished! Disabling...");
			getServer().getPluginManager().disablePlugin(this);
		}
	}

	@Override
	public void onDisable() {
		configHandler.load();
		configHandler.addComments();
		info("GetPerms disabled");
	}

	private final void debugValues() {
		String cfgV = pluginVersion;
		boolean firstRun = configuration.getBoolean("firstRun", false);
		boolean sendStats = configuration.getBoolean("sendStats", true);
		boolean silentMode = configuration.getBoolean("silentMode", false);
		boolean autoGenerate = configuration.getBoolean("autoGenerate", true);
		boolean regenerateOnPluginChange = configuration.getBoolean("regenerateOnPluginChange", true);
		boolean autoUpdate = configuration.getBoolean("autoUpdate", true);
		boolean autoDownload = configuration.getBoolean("autoDownload", true);
		boolean disableOnFinish = configuration.getBoolean("disableOnFinish", false);
		boolean devBuilds = configuration.getBoolean("devBuilds", false);
		boolean debugMode = configuration.getBoolean("debugMode", false);

		config("cfgV: \"" + cfgV + "\"");
		config("firstRun: " + firstRun);
		config("sendStats: " + sendStats);
		config("silentMode: " + silentMode);
		config("autoGenerate: " + autoGenerate);
		config("regenerateOnPluginChange: " + regenerateOnPluginChange);
		config("autoUpdate: " + autoUpdate);
		config("autoDownload: " + autoDownload);
		config("disableOnFinish: " + disableOnFinish);
		config("devBuilds: " + devBuilds);
		config("debugMode: " + debugMode);
	}

	public void runUpdateTask() {
		scheduler.scheduleSyncDelayedTask(this, new UpdateTask(this), 10L);
	}

	public final void printStackTrace(Exception e) {
		if (!configuration.getBoolean("silentMode", false)) {
			severe("An error has occurred! Please file an issue at");
			severe("https://github.com/GetPerms/GetPerms/issues with the");
			severe("below stack trace. Also, please include details about");
			severe("what caused this error to occur.");
			e.printStackTrace();
		}
	}

	public final void generateFiles(boolean internal) {
		// internal is true if called by onEnable and False if called by command
		if (internal && !generationFinished) {
			// 5 * 20 = 5 seconds (20 ticks per second)
			scheduler.scheduleSyncDelayedTask(this, new PermissionTask(this), 5 * 20L);
		} else if (!internal) {
			scheduler.scheduleSyncDelayedTask(this, new PermissionTask(this));
		}
	}

	// Gets the readme and changelog
	private void getStartFiles() {
		try {
			dataFolder.mkdir();
			info("Downloading changelog and readme...");
			downloadFile("https://raw.github.com/GetPerms/GetPerms/master/Changelog.txt", changeLog);
			downloadFile("https://raw.github.com/GetPerms/GetPerms/master/ReadMe.txt", readMe);
			info("The changelog and readme can be found in 'plugins/GetPerms/'");
			debug("Downloads succeded. firstRun being set to false...");
			configuration.set("firstRun", false);
		} catch (MalformedURLException e) {
			debug("MalformedURLException thrown, setting firstRun to true...");
			configuration.set("firstRun", true);
			printStackTrace(e);
			warn("Error downloading readme and changelog!");
			info("The readme is available at");
			info("https://raw.github.com/GetPerms/GetPerms/master/ReadMe.txt");
			info("and the changelog is available at");
			info("https://raw.github.com/GetPerms/GetPerms/master/Changelog.txt");
		} catch (FileNotFoundException e) {
			debug("FileNotFoundException thrown, setting firstRun to true...");
			configuration.set("firstRun", true);
			printStackTrace(e);
			warn("Error downloading readme and changelog!");
			info("The readme is available at");
			info("https://raw.github.com/GetPerms/GetPerms/master/ReadMe.txt");
			info("and the changelog is available at");
			info("https://raw.github.com/GetPerms/GetPerms/master/Changelog.txt");
		} catch (IOException e) {
			debug("IOException thrown, setting firstRun to true...");
			configuration.set("firstRun", true);
			printStackTrace(e);
			warn("Error downloading readme and changelog!");
			info("The readme is available at");
			info("https://raw.github.com/GetPerms/GetPerms/master/ReadMe.txt");
			info("and the changelog is available at");
			info("https://raw.github.com/GetPerms/GetPerms/master/Changelog.txt");
		} catch (NoSuchAlgorithmException e) {
			debug("NoSuchAlgorithmException thrown, setting firstRun to true...");
			configuration.set("firstRun", true);
			printStackTrace(e);
			warn("Error downloading readme and changelog!");
			info("The readme is available at");
			info("https://raw.github.com/GetPerms/GetPerms/master/ReadMe.txt");
			info("and the changelog is available at");
			info("https://raw.github.com/GetPerms/GetPerms/master/Changelog.txt");
		}
	}

	public String downloadFile(String url, File file)
			throws MalformedURLException, IOException, NoSuchAlgorithmException {
		MessageDigest messageDigest = MessageDigest.getInstance("MD5");

		DigestInputStream inputStream = new DigestInputStream(
				new BufferedInputStream(new java.net.URL(url).openStream()), messageDigest);
		FileOutputStream fileOutputStream = new FileOutputStream(file);
		BufferedOutputStream bufferedOutputStream = new BufferedOutputStream(fileOutputStream, 1024);
		byte[] data = new byte[1024];
		int i = 0;
		while ((i = inputStream.read(data, 0, 1024)) >= 0) {
			bufferedOutputStream.write(data, 0, i);
		}
		bufferedOutputStream.close();
		inputStream.close();

		return convertByteArrayToHexString(messageDigest.digest());
	}

	private String convertByteArrayToHexString(byte[] arrayBytes) {
		StringBuffer stringBuffer = new StringBuffer();
		for (byte arrayByte : arrayBytes) {
			stringBuffer.append(Integer.toString((arrayByte & 0xff) + 0x100, 16).substring(1));
		}
		return stringBuffer.toString();
	}

	private boolean comparePlugins() {
		boolean ok = true;
		Plugin[] pluginList = pluginManager.getPlugins();
		Configuration pluginListConfiguration = pluginListHandler.getConfig();

		for (Plugin plugin : pluginList) {
			if (pluginListConfiguration.contains(plugin.getDescription().getName())) {
				if (!pluginListConfiguration.getString(plugin.getDescription().getName())
						.equals(plugin.getDescription().getVersion())) {
					ok = false;
					break;
				}
			} else {
				ok = false;
				break;
			}
		}

		return ok;
	}

	private void writePluginList() {
		Plugin[] pluginList = pluginManager.getPlugins();
		Configuration pluginListConfiguration = pluginListHandler.getConfig();
		Set<String> keys = pluginListConfiguration.getKeys(false);

		// Clear the configuration
		for (String key : keys) {
			pluginListConfiguration.set(key, null);
		}

		for (Plugin plugin : pluginList) {
			pluginListConfiguration.set(plugin.getDescription().getName(), plugin.getDescription().getVersion());
		}

		pluginListHandler.save();
	}

	public void registerUpdateTask(int taskId) {
		if (updateTaskId != -1) {
			Bukkit.getScheduler().cancelTask(updateTaskId);
		}

		updateTaskId = taskId;
	}

	public void debug(String i) {
		if (configuration.getBoolean("debugMode", false)) {
			if (!configuration.getBoolean("silentMode", false)) {
				logger.info("[Debug] " + i);
			}
		}
	}

	public void config(String i) {
		if (configuration.getBoolean("debugMode", false)) {
			if (!configuration.getBoolean("silentMode", false)) {
				logger.config("[Debug] " + i);
			}
		}
	}

	public void info(String i) {
		if (!configuration.getBoolean("silentMode", false)) {
			logger.info(i);
		}
	}

	public void warn(String i) {
		if (!configuration.getBoolean("silentMode", false)) {
			logger.warning(i);
		}
	}

	public void severe(String i) {
		if (!configuration.getBoolean("silentMode", false)) {
			logger.severe(i);
		}
	}

	private final void cleanOldFiles() {
		info("Cleaning up old files from previous versions...");

		File[] oldFileList = {new File(dataFolder, "pluginlist.txt"), new File("pnodes.txt"),
				new File("pnodesfull.txt"), new File("EssentialsPnodes.txt"), new File("EssentialsPnodesfull.txt"),
				new File("update/GetPerms.jar"), new File("update/GPtemp.jar"),
				new File("Essentials_permission_nodes.txt"), new File("Essentials_permission_nodes_desc.txt"),
				new File(dataFolder, "permission_nodes.txt"), new File(dataFolder, "permission_nodes_desc.txt"),
				new File(dataFolder, "EssentialsPnodes.txt"), new File(dataFolder, "EssentialsPnodesfull.txt")};
		File[] oldSoftDirectoryList = {new File("update")};

		for (File file : oldFileList) {
			if (file.exists()) {
				file.delete();
			}
		}

		for (File file : oldSoftDirectoryList) {
			if (file.exists() && file.isDirectory()) {
				if (file.listFiles().length == 0) {
					file.delete();
				}
			}
		}
	}
}